global/scripts/debug.shlib

```diff
@@ -0,0 +1,45 @@
+# Description:
+#
+#   Function for printing debugging information on caught errors.
+#   It makes use of error trapping. To reset the trap, add
+#   end of the script using "trap - EXIT".
+#
+# Input:
+#
+#   — line number
+#   - command ran
+#   - exit status
+
+trap 'on_error $LINENO "$BASH_COMMAND" $?; exit' ERR
+
+on_error()
+{
+    me=`basename $0`
+    line=($2)
+    com=${line[0]}
+    arg=${line[@]:1}
+    echo
+    echo -e "\033[33m##### ERROR: Execution stopped prematurely #####"
```

Comment:
Error messages should probably be printed to standard error, not standard out, and I'd prefer not terminal putting color codes into log files.

Reply:
I asked @tbbrown about this once, and he made a reasonable case for why the error messages should at least go to both stdout and stderr -- namely, people are more likely to look at the end of stdout.

Reply:
I edited our log_Err* functions so that they always put something in both stderr and stdout.  This will result in some redundant messages when running existing scripts with parameters missing in a terminal (no log redirection), because currently log_Err duplicates the entire message (while log_Err_Abort uses a generic message on stdout).  A good middle ground may be to put the detailed info only into stderr, as long as both stderr and stdout indicate that something is wrong.

Another option is to actually have the log_Err functions check if stderr is connected to a tty, and change whether they duplicate messages based on that.

Reply:
Checking whether stderr is connected to a tty, and modifying the behavior based on that, would be a neat way to avoid redundant messages to the terminal.

Reply:
I have added this tty check.  I doubt there is a cross-platform way to check for stderr and stdout both being redirected to the same file (mac doesn't have /proc), so that case will have duplicate messages, but avoiding duplicates with no redirection seems good enough.